src/mme-app/actionHandlers/s1HandoverActionHandlers.cpp

```diff
@@ -0,0 +1,659 @@
+    return ActStatus::PROCEED;
+}
+
+/***************************************
+* Action handler : send_s1_rel_cmd_to_ue_for_ho
+***************************************/
+ActStatus ActionHandlers::send_s1_rel_cmd_to_src_enb_for_ho(ControlBlock& cb)
+{
+    log_msg(LOG_DEBUG, "Inside send_s1_rel_cmd_to_ue_for_ho\n");
+    UEContext *ue_ctxt = static_cast<UEContext*>(cb.getPermDataBlock());
+    if(ue_ctxt == NULL)
+    {
+    	log_msg(LOG_DEBUG, "send_s1_rel_cmd_to_ue_for_ho: ue context is NULL \n");
+    	return ActStatus::HALT;
+    }
+
+    S1HandoverProcedureContext *ho_ctxt =
+            dynamic_cast<S1HandoverProcedureContext*>(cb.getTempDataBlock());
+    if (ho_ctxt == NULL)
+    {
+        log_msg(LOG_DEBUG, "send_s1_rel_cmd_to_ue_for_ho: procedure ctxt is NULL \n");
+        return ActStatus::HALT;
+    }
+    struct s1relcmd_info s1relcmd;
+    s1relcmd.msg_type = s1_release_command;
+    s1relcmd.ue_idx = ue_ctxt->getContextID();
+    s1relcmd.enb_s1ap_ue_id = ho_ctxt->getSrcS1apEnbUeId();
+    s1relcmd.cause.present = s1apCause_PR_radioNetwork;
+    s1relcmd.cause.choice.radioNetwork = s1apCauseRadioNetwork_successful_handover;
+    s1relcmd.enb_fd = ho_ctxt->getSrcEnbContextId();
+
+    // Fire and forget s1 release to src enb
+
+    /*Send message to S1AP-APP*/
+    cmn::ipc::IpcAddress destAddr = {TipcServiceInstance::s1apAppInstanceNum_c};
+    MmeIpcInterface &mmeIpcIf = static_cast<MmeIpcInterface&>(compDb.getComponent(MmeIpcInterfaceCompId));
+    mmeIpcIf.dispatchIpcMsg((char *) &s1relcmd, sizeof(s1relcmd), destAddr);
+
+    ProcedureStats::num_of_s1_rel_cmd_sent ++;
+    log_msg(LOG_DEBUG,"Leaving send_s1ap_ue_ctxt_rel_command_to_src_enb \n");
+    return ActStatus::PROCEED;
+}
+
+/***************************************
+ * Action handler : ho_complete
+ ***************************************/
+ActStatus ActionHandlers::ho_complete(ControlBlock &cb)
+{
+    log_msg(LOG_INFO, "Inside ho_complete\n");
+
+    ProcedureStats::num_of_ho_complete++;
+
+    MmeContextManagerUtils::deallocateProcedureCtxt(cb, s1Handover_c);
+
+    return ActStatus::PROCEED;
+}
+
+/***************************************
+* Action handler : is_tau_required
+***************************************/
+ActStatus ActionHandlers::is_tau_required(ControlBlock& cb)
+{
+    log_msg(LOG_INFO, "Inside is_tau_required\n");
+    UEContext *ue_ctxt = static_cast<UEContext*>(cb.getPermDataBlock());
+    if (ue_ctxt == NULL)
+    {
+        log_msg(LOG_ERROR, "is_tau_required: ue context is NULL\n",
+                cb.getCBIndex());
+        return ActStatus::HALT;
+    }
+
+    S1HandoverProcedureContext *s1HoPrcdCtxt_p =
+            dynamic_cast<S1HandoverProcedureContext*>(cb.getTempDataBlock());
+    if (s1HoPrcdCtxt_p == NULL)
+    {
+        log_msg(LOG_DEBUG,
+                "is_tau_required: S1HandoverProcedureContext is NULL\n");
+        return ActStatus::HALT;
+    }
+
+    if (ue_ctxt->getTai() == s1HoPrcdCtxt_p->getTargetTai())
```

Comment:
We already updated TAI before sending MBReq...in that case this comparison will always be true...isn't it ?

Reply:
TAI is updated to UE Context only at the end of TAU procedure. The TAI in UE Context is still the old one.